src/firmware/linux/host/types.rs

```diff
@@ -250,3 +255,218 @@ impl SnpConfig {
+    ///
+    /// ```C
+    /// struct cert_table {
+    ///    struct {
+    ///       unsigned char guid[16];
+    ///       uint32 offset;
+    ///       uint32 length;
+    ///    } cert_table_entry[];
+    /// };
+    /// ```
+    ///
+    unsafe fn parse_table(&self) -> Vec<UapiCertTableEntry> {
+        const OFFSET_START: isize = (std::mem::size_of::<u8>() * 16) as isize;
+        const LENGTH_START: isize = (std::mem::size_of::<u8>() * 20) as isize;
+        const ENTRY_SIZE: isize = (std::mem::size_of::<u8>() * 24) as isize;
+        const GUID_SIZE: usize = 16_usize;
+
+        let mut retval: Vec<UapiCertTableEntry> = vec![];
+        let mut guid_ptr: *const CertTableEntry = self.entries;
+        let mut guid: String;
+        let mut cert_address: *mut u8;
+        let mut cert_len: usize;
+        let mut cert_end: *mut u8;
+
+        loop {
+            let mut cert_bytes: Vec<u8> = vec![];
+
+            // Calculate values for each CertTableEntry.
+            cert_address = guid_ptr.offset(OFFSET_START) as *mut u8;
+            cert_len = *(guid_ptr.offset(LENGTH_START) as *const usize);
+            cert_end = cert_address.add(cert_len);
+            guid = String::from_raw_parts(guid_ptr as *mut u8, GUID_SIZE, GUID_SIZE);
+
+            // Gather the certificate bytes.
+            while cert_address.ne(&cert_end) {
+                cert_bytes.push(*cert_address);
+                cert_address = cert_address.add(1);
+            }
+
+            // Append the entry to the vector.
+            retval.push(UapiCertTableEntry::from_guid(&guid, cert_bytes));
+
+            // Move the pointer ahead to the next value.
+            guid_ptr = guid_ptr.offset(ENTRY_SIZE);
+
+            // Check for the end of the table.
+            if guid_ptr.eq(&std::ptr::null()) {
+                break;
+            }
+        }
+
+        retval
+    }
+
+    pub fn to_uapi(&self) -> UapiCertTable {
+        UapiCertTable {
+            entries: unsafe { self.parse_table() },
+        }
+    }
+}
+
+#[derive(Default)]
+#[repr(C)]
+pub struct SnpSetExtConfig {
+    /// Address of the SnpConfig or 0 when reported_tcb does not need
+    /// to be updated.
+    pub config_address: u64,
+
+    /// Address of extended guest request [`CertTable`] or 0 when
+    /// previous certificate should be removed on SNP_SET_EXT_CONFIG.
+    pub certs_address: u64,
+
+    /// Length of the certificates.
+    pub certs_len: u32,
+}
+
+impl SnpSetExtConfig {
+    pub(crate) fn from_uapi(data: &SnpExtConfig) -> Self {
+        Self {
+            certs_address: if data.certs.is_none() {
+                0
+            } else {
+                &CertTable::from_uapi(data.certs.clone().unwrap()) as *const CertTable as u64
+            },
+            config_address: if data.config.is_none() {
+                0
+            } else {
+                &data.config.unwrap() as *const SnpConfig as u64
+            },
+            certs_len: data.certs_len,
```

Comment:
Do you think it's necessary to add a check, if `certs_address.is_none()` then `certs_len` should be 0?

Reply:
I changed it to:

```rust
certs_len: if data.certs.is_none() {
    0
} else {
    data.certs_len
},
```